middleman_protocol/middleman_protocol/helpers.py

```diff
@@ -0,0 +1,81 @@
+import socket
+
+from mypy.types import Iterable
+
+from middleman_protocol.constants import ESCAPE_CHARACTER
+from middleman_protocol.constants import ESCAPE_SEQUENCES
+from middleman_protocol.constants import FRAME_PAYLOAD_LENGTH_STARTING_BYTE
+from middleman_protocol.constants import FRAME_PAYLOAD_STARTING_BYTE
+from middleman_protocol.constants import FRAME_SEPARATOR
+from middleman_protocol.constants import FRAME_SEPARATOR_LENGTH
+from middleman_protocol.constants import RECEIVE_BYTES_PER_LOOP
+from middleman_protocol.message import AbstractMiddlemanMessage
+
+
+def handle_message_receive(connection: socket.socket, encoding_expected: bool=False) -> Iterable[bytes]:
+    """
+    Helper generator for receiving serialized Middleman protocol messages.
+
+    Yields received bytes.
+    """
+    assert isinstance(connection, socket.socket)
+    assert isinstance(encoding_expected, bool)
+
+    received_data = b''
+    message_with_payload_length = None
+    # Used to mark how many bytes was received since last decoding, to avoid decoding twice same part of data.
+    decoding_buffer = 0
+
+    while True:
+        received_data += connection.recv(RECEIVE_BYTES_PER_LOOP)
+
+        # If encoding is expected to be used by data sender,
+        # and there is encoded sequence in last received data,
+        # decode it and set encoding buffer.
+        if (
+            encoding_expected is True and (
+                ESCAPE_SEQUENCES[FRAME_SEPARATOR] in received_data[decoding_buffer:] or
+                ESCAPE_SEQUENCES[ESCAPE_CHARACTER] in received_data[decoding_buffer:]
+            )
+        ):
+            received_data = (
+                received_data[:decoding_buffer] +
+                AbstractMiddlemanMessage.escape_decode_raw_message(received_data[decoding_buffer:])
+            )
+            decoding_buffer = len(received_data)
+
+        # If separator is met, reset received data and restart.
+        if (
+            len(received_data) >= FRAME_SEPARATOR_LENGTH and
+            received_data[-FRAME_SEPARATOR_LENGTH:] == FRAME_SEPARATOR
+        ):
+            received_data = b''
+            message_with_payload_length = None
+            decoding_buffer = 0
+            continue
+
+        # When defined amount of bytes is received, read payload length.
+        # There is a range defined, because it might happen that payload length matches part of one of escape sequences,
+        # so loop should read it completely first, then decode, and then finally read payload length.
+        if FRAME_PAYLOAD_STARTING_BYTE <= len(received_data) <= (FRAME_PAYLOAD_STARTING_BYTE + len(ESCAPE_CHARACTER)):
+            # Last received data must not match start of ESCAPE_CHARACTER,
+            # otherwise ESCAPE_CHARACTER should be read completely before reading payload length.
```

Comment:
@cameel 
Well, it started with something that was supposed to be a simple handler, but ended up in being more complex task then the original one. It took most of my time last week to identify the edge cases and implement solutions for them. What is bugging me is that those edge cases, as the need for encoding itself, will happen very very rarely.

> What if the second half of one of the escape sequences is the same as the first half of the separator so that when you combine the escape sequence with some random bytes you suddenly get a separator?

I think this can be achieved easily if both escape sequences and separator will not be completely random, but predefined in a way that this can't happen. Even better, we can do that escape sequences will be generated (in some kind of transform function) from separator. This way we can still use random separator, and be sure that the situation above will not happen (it can even be Caesar's algorithm).

> What if payload_length is inside the escape sequence and that sequence is immediately followed by a separator?

I think this case is already covered, I will test it as soon as I will finish writing this message.

> But for now I'll say that I suspect that it may be impossible to get it right with multi-character separators and escape sequences without also making it very complex and convoluted.

Nobody said it is going to be easy ;) But I think most of the work is done, and most edge cases are already covered, and we still can make something out of it. 

What I suggest to do now is
1. You can think of other edge cases (I will do that too, however I thought I already have everything covered).
2. Write them down, I will take a look, we can discuss.
3. I will implement solutions and write tests.

I really think the code is not that much complicated right now, and adding solutions for any new edge case found will not make it much more complicated.

Reply:
@cameel 

>> What if payload_length is inside the escape sequence and that sequence is immediately followed by a separator?
> I think this case is already covered, I will test it as soon as I will finish writing this message.

I have added three tests for it.
1. Case 1, payload_length starts with the start of escape sequence and that sequence is immediately followed by a separator.
2. Case 2, payload_length ends with the end of escape sequence and that sequence is immediately followed by a separator.
3. Case 3, payload_length is in the middle of escape sequence and that sequence is immediately followed by a separator.

Along with small fix in helper, (Which was there before, but I thought it was finally not necessary, and remove it. Looks like after all it is needed.) they all pass.

You can check the fixups directly:
5a7f51160acdc871ff6d491e08da9d3a5f3170cb
8ed6143cf76d006093c6557c4cacfb9a56cf778d

Reply:
> I think this can be achieved easily if both escape sequences and separator will not be completely random, but predefined in a way that this can't happen. Even better, we can do that escape sequences will be generated (in some kind of transform function) from separator. This way we can still use random separator, and be sure that the situation above will not happen (it can even be Caesar's algorithm).

The separator is not random. It's just chosen based random input. https://xkcd.com/221/ ;) It's hardcoded so it's always the same. I'm mostly worried about the case in which someone decides that the current separator has some issue and changes it to something else that does not meet all conditions required to prevent weird edge cases. But that's what assertions are for. They're there to tell someone reading it what we expect. We just need more of them.

First of all I think that requiring that no character used in separator is ever used within the escape character would prevent a lot of edge cases. Please add an assert for that in `constants.py`:

``` python
assert set(ESCAPE_CHARACTER) & set(FRAME_SEPARATOR) == set()
```

That's not enough though - I'll follow up with more edge cases. The ones I'm most worried about are where the message ends with a fragment of an escape character and we have to wait for more. They don't break the escaping but make us wait for the next message which is something I want to avoid.

As for generating escape sequences from separator using a formula - that's a good idea. But the assertions are needed anyway.

Reply:
I have a case where the generator will not return a message until the client starts sending another one. And if the client closes the connection instead of sending another one, the message will not be returned from your generator at all.

Let's assume that `payload_length` has one byte and starts at position 2 (counting from 0).

|                          |               |
|--------------------------|---------------|
| Separator                | `66 66 66`
| Escape character         | `33 01 33 33`
| Escaped separator        | `33 01 33 33 00`
| Escaped escape character | `33 01 33 33 01`
|                          |               |
| Input                    | `22 33 01 33`

That's a complete message and `33 01 33` is not a part of an escape sequence but you don't have any way to know that. It's just a coincidence that it starts with the same characters.

You won't know that until you see a separator after it. But the client might just close the connection and not send anything. Or it may send a separator but, say, a minute later when it starts sending a new message.

The only way to avoid this I see now would be to escape every prefix of the escape character too.

Reply:
Let's add
``` python
assert len(ESCAPE_CHARACTER) == 1
```

Since we have to escape every prefix, including the first character, there's no point in making it any longer than one character. The problem with using only one is that it will make messages 1/256 bytes longer on average compared to escaping only one multi-character sequence (which would be much rarer). But it's still less often than if we escaped every prefix of such a multi-character sequence.

Reply:
OK. So, as indicated in https://github.com/golemfactory/concent/pull/673/files#r206841640 we have to switch to single-character separator and escape character. Escape sequences can have more than two characters but we don't have a need for it at this moment so let's make them all two characters long.

With that you can simplify the code and split it into two nested generators as I suggested above (names are just my suggestions):
#### `unescape_stream()`
- While there's another byte
    1. If it's the separator
        - If the flag indicating that we're waiting for the second character of an escape sequence is set
            - yield special value indicating an invalid frame
        - Reset the "frame invalid" flag
        - yield special value indicating an end of a frame
    2. If you were waiting for the second part of an escape sequence
        - If this is not a valid escape sequence
            - mark frame as invalid
            - yield special value indicating an invalid frame
            - continue
        - Otherwise
            - unescape the sequence
            - yield the result
            - continue
    3. If the frame is marked as invalid
        - continue
    4. If it's an escape character
        - set a flag indicating that you're waiting for the other byte
        - continue
    5. yield the current byte.

#### `receive_frame()`
- While `unescape_stream()` yields another byte:
    1. If you have just received the value indicating the end of a frame
        - If you have not yielded this frame yet
            - Yield the content of the data buffer
            - Clear the data buffer
        - Clear the flag indicating that you have yielded this frame
        - Clear stored payload length
        - continue
    2. If you have just received the value indicating an invalid frame
        - yield a special value indicating an invalid frame
        - continue
    3. If you have not yielded this frame yet
        1. If you have just received the last byte of `payload_length`
            - Store the length.
        2. If you have just received the last byte of the payload (which includes the situation where payload is empty)
            - Yield the content of the data buffer
            - Clear the data buffer
            - Set the flag indicating that you have yielded this frame

Reply:
During the discussion with @dybi and @rwrzesien we have decided to make additional assumptions to simplify the implementation:
1. Both separator and the escape character must be 1-byte long.
2. `payload_length` will be removed. Now all the bytes between the header and the next separator are considered a part of a frame.
    - End of stream counts as a separator.
3. Due to 2. we always wait for the separator. If the client wants the message to be processed immediately it should send the separator immediately after it.

I have updated the pseudocode above with that assumptions. Basically, `unescape_stream()` did not change at all, `receive_frame()` is a bit simpler.

Reply:
This can actually be simplified even more. In this version `receive_frame()` returns frame class instances:

#### `split_stream()`
- While there's another byte
    1. If it's the separator or the connection has been closed
        - yield the data buffer
    2. Otherwise
        - Add the character to the buffer

As a bonus this also solves #684.

#### `unescape_stream()`
- for `escaped_frame` in `unescape_stream()`
    - For every occurrence of the escape character in `escaped_frame`:
        - If it's the last character in the frame
            - yield special value indicating a frame with broken escaping
            - continue the outer loop
        - Otherwise if this character combined with the next one is not a valid escape sequence
            - yield special value indicating a frame with broken escaping
            - continue the outer loop
        - Otherwise
            - Replace the escape sequence with the character it translates to
    - yield unescaped frame

#### `receive_frame()`
- for `unescaped_frame` in `unescape_stream()`
    - if it's the special value indicating a frame with broken escaping
        - yield it
    - otherwise
        - create an instance of the frame class
        - yield the instance

Reply:
:+1: done

However I don't see how it should resolve #684. If for example you change amount of received bytes to 1000, and it that part you will receive two frame separators, it won't even return the data.